lib/msal-core/src/AuthError.ts

```diff
@@ -0,0 +1,240 @@
+ * Error thrown when there is an error in the client code running on the browser.
+ */
+export class ClientAuthError extends AuthError {
+    constructor(error: string, errorDesc: string) {
+        super(error, errorDesc);
+        this.name = "ClientAuthError";
+
+        Object.setPrototypeOf(this, ClientAuthError.prototype);
+    }
+
+    static createEndpointResolutionError(errDesc: string): ClientAuthError {
+        var errorMessage = "Error: could not resolve endpoints. Please check network and try again.";
+        if (!Utils.isEmpty(errDesc)) {
+            errorMessage += " Details: " + errDesc;
+        }
+        return new ClientAuthError(ErrorMessage.endpointResolutionError.code, errorMessage);
+    }
+
+    static createMultipleMatchingTokensInCacheError(scope: string): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.multipleMatchingTokens.code,
+            "Cache error for scope " + scope + ": " + ErrorMessage.multipleMatchingTokens.desc);
+    }
+
+    static createMultipleAuthoritiesInCacheError(scope: string): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.multipleMatchingAuthorities.code,
+            "Cache error for scope " + scope + ": " + ErrorMessage.multipleMatchingAuthorities.desc);
+    }
+
+    static createPopupWindowError(): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.popUpWindowError.code,
+            ErrorMessage.popUpWindowError.desc);
+    }
+
+    static createTokenRenewalTimeoutError(): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.tokenRenewalError.code,
+            ErrorMessage.tokenRenewalError.desc);
+    }
+
+    static createInvalidStateError(invalidState: string, actualState: string): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.invalidStateError.code,
+            ErrorMessage.invalidStateError.desc + invalidState + ", state expected : " + actualState);
+    }
+
+    static createNonceMismatchError(invalidNonce: string, actualNonce: string): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.nonceMismatchError.code,
+            ErrorMessage.nonceMismatchError + invalidNonce + ", nonce expected : " + actualNonce);
+    }
+
+    static createLoginInProgressError(): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.loginProgressError.code,
+            ErrorMessage.loginProgressError.desc);
+    }
+
+    static createAcquireTokenInProgressError(): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.acquireTokenProgressError.code,
+            ErrorMessage.acquireTokenProgressError.desc);
+    }
+
+    static createUserCancelledError(): ClientAuthError {
+        return new ClientAuthError(ErrorMessage.userCancelledError.code,
+            ErrorMessage.userCancelledError.desc);
+    }
+}
+
+/**
+ * @hidden
+ *
+ * Error thrown when there is an error in configuration of the .js library.
+ */
+export class ConfigurationError extends ClientAuthError {
```

Comment:
Should this extend AuthError not ClientAuth Error?

Reply:
It was agreed we would have AuthError as the base, ClientAuthError, ServerError extending this base class, and other error types deriving from these two types.

Reply:
I do remember that but there is no benefit out of that right? We don't represent in an way to developer.

You dont get an error with any indication that a config error is a client error right ?

Reply:
ConfigError instanceof ClientAuthError = false
ClientAuthError instanceof ConfigError = true

usage:

try {
msal.loginRedirect(...);
} catch (err) {
    if (err instanceof ClientAuthError) {
         if(err instanceof ConfigError) {

         }
    }
}

The benefit is that errors are classified as coming from the client or server before being classified as configuration or client library errors. Happy to discuss changing these, but I will keep it as is for now.

Reply:
that makes sense, I am fine with that.